### ISSUE TITLE
unsupported tensor op : InceptionResnetV1/Conv2d_1a_3x3/BatchNorm/cond/AssignMovingAvg

### ISSUE THREAD

**Issue by solderzzc** [ARM-software/armnn#15]
I'm converting our tensorflow pb file into armnn, I got following exception with the latest code:
```
terminate called after throwing an instance of 'armnn::ParseException'
  what():  Node 'InceptionResnetV1/Conv2d_1a_3x3/BatchNorm/cond/Identity_1' has Control Input '^InceptionResnetV1/Conv2d_1a_3x3/BatchNorm/cond/AssignMovingAvg' which is unsupported.
```

**Comment by MatthewARM** [comment]
Hi @solderzzc the current tensorflow parser cannot guarantee the correct operation of control inputs (which is tensorflow's way of guaranteeing a particular order of execution).

You can try removing the code block at TfParser.cpp:489 to get past this parse error, but if the control inputs are guaranteeing execution of some variable assignment or something, you will probably just hit another parse error. The full list of supported TensorFlow ops is at: src/armnnTfParser/TensorFlowSupport.md

It seems to me that maybe it's not a frozen inference-only model? If you try following the steps at https://www.tensorflow.org/mobile/prepare_models it will remove all training-only nodes and you will have a much smaller, cleaner graph that is more likely to work with ArmNN.

All the best,
Matthew

**Comment by solderzzc** [comment]
Hi @MatthewARM 

Thank for your reply. I removed the code block at TfParser.cpp to bypass parser error. The model has `phase_train` (ref: https://github.com/davidsandberg/facenet/issues/357#issuecomment-315548160) to control graph inference flow, so it might has issue during runtime.

I also called `optimize_for_inference` API of tensorflow to have `training-only` nodes removed(same behavior as mentioned in https://www.tensorflow.org/mobile/prepare_models), then hit another error log:

```
terminate called after throwing an instance of 'armnn::ParseException'
  what():  Currently only FLOAT is supported for tensorflow nodes (apart from Const)
Aborted
```

I think this is because that FLOAT variables were converted to CONST during `Freeze Graph` phase.

Reference:
https://github.com/davidsandberg/facenet/blob/master/src/freeze_graph.py#L88 

```
graph_util.convert_variables_to_constants(
        sess, input_graph_def, output_node_names.split(","),
        variable_names_whitelist=whitelist_names)
```

Thanks
Simba

**Comment by MatthewARM** [comment]
Hi @solderzzc "Currently only FLOAT is supported for tensorflow nodes (apart from Const)" indicates that there is an integer operation in your network, do you know if that is the case? Currently we only support float operations.

If you can find out which operation, and what data type, it would help us prioritise the support work internally. Or we can find a workaround.

All the best,
Matthew

**Comment by solderzzc** [comment]
Hi, @MatthewARM 
I just go through the network design of facenet,  the `non float` [ops](https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py#L238) should come from `dropout` layer of [tensorflow.slim](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L1579).

### Network define with dropout
![screen shot 2018-06-08 at 10 41 19 am](https://user-images.githubusercontent.com/3085564/41172413-a941224a-6b08-11e8-96a8-5c27cd85df04.png)

### Slim dropout API
![screen shot 2018-06-08 at 10 36 03 am](https://user-images.githubusercontent.com/3085564/41172144-c877660c-6b07-11e8-9584-f03bd0e371eb.png)

Best Regards
Simba

**Comment by MatthewARM** [comment]
Hi @solderzzc, dropout should definitely get removed by the 'strip unused nodes' transform. Maybe you could try the mobile inference preparation steps again?

**Comment by solderzzc** [comment]
Hi @MatthewARM, I loaded the graph into tensorboard, the 'Dropout' layer is not included in the freezed graph. The issue should be introduced by `DT_BOOL` `phase_train`.
![screen shot 2018-06-11 at 12 01 37 pm](https://user-images.githubusercontent.com/3085564/41251718-a73eaa9e-6d6f-11e8-923e-d9d8c929b002.png)
![screen shot 2018-06-11 at 12 02 25 pm](https://user-images.githubusercontent.com/3085564/41251720-a75945ac-6d6f-11e8-932c-d9f5529079c1.png)
![screen shot 2018-06-11 at 12 02 55 pm](https://user-images.githubusercontent.com/3085564/41251722-a781579a-6d6f-11e8-93cc-815998246c4c.png)

**Comment by MatthewARM** [comment]
Thanks for the detailed analysis, @solderzzc. 

We can try to get a proper fix done in a future release, but I think you can work around the issue by removing all connections to 'phase_train' from your graph. In the ArmNN TensorFlow parser we will always assume that 'phase_train' is FALSE.

**Comment by solderzzc** [comment]
@MatthewARM I will try it and look forward for the next release.

Thanks
